test/functional/rpc_getblockstats.py

```diff
@@ -0,0 +1,180 @@
+        assert_equal(stats[1]['avgfeerate'], 20)
+        # assert_equal(stats[1]['mintxsize'], 192)
+        # assert_equal(stats[1]['maxtxsize'], 192)
+        # assert_equal(stats[1]['mediantxsize'], 192)
+        # assert_equal(stats[1]['avgtxsize'], 192)
+
+        assert_equal(stats[max_stat_pos]['txs'], 4)
+        assert_equal(stats[max_stat_pos]['swtxs'], 2)
+        assert_equal(stats[max_stat_pos]['ins'], 3)
+        assert_equal(stats[max_stat_pos]['outs'], 8)
+        assert_equal(stats[max_stat_pos]['totalfee'], 56880)
+        assert_equal(stats[max_stat_pos]['utxo_increase'], 5)
+        assert_equal(stats[max_stat_pos]['utxo_size_inc'], 380)
+        # assert_equal(stats[max_stat_pos]['total_size'], 643)
+        # assert_equal(stats[max_stat_pos]['total_weight'], 2572)
+        assert_equal(stats[max_stat_pos]['total_out'], 9999939360)
+        assert_equal(stats[max_stat_pos]['minfee'], 3320)
+        assert_equal(stats[max_stat_pos]['maxfee'], 49800)
+        assert_equal(stats[max_stat_pos]['medianfee'], 3760)
+        assert_equal(stats[max_stat_pos]['avgfee'], 18960)
+        assert_equal(stats[max_stat_pos]['minfeerate'], 20)
+        # assert_equal(stats[max_stat_pos]['maxfeerate'], 300)
+        assert_equal(stats[max_stat_pos]['medianfeerate'], 20)
+        assert_equal(stats[max_stat_pos]['avgfeerate'], 109)
+        # assert_equal(stats[max_stat_pos]['mintxsize'], 192)
+        # assert_equal(stats[max_stat_pos]['maxtxsize'], 226)
+        # assert_equal(stats[max_stat_pos]['mediantxsize'], 225)
+        # assert_equal(stats[max_stat_pos]['avgtxsize'], 214)
+
+        # Test invalid parameters raise the proper json exceptions
+        tip = start_height + max_stat_pos
+        assert_raises_rpc_error(-8, 'Target block height %d after current tip %d' % (tip+1, tip), node.getblockstats, height=tip+1)
+        assert_raises_rpc_error(-8, 'Target block height %d is negative' % (-1), node.getblockstats, height=-tip-2)
+
+        # Make sure not valid stats aren't allowed
+        inv_sel_stat = 'asdfghjkl'
+        inv_stats = [
+            [inv_sel_stat],
+            ['minfee' , inv_sel_stat],
+            [inv_sel_stat, 'minfee'],
+            ['minfee', inv_sel_stat, 'maxfee'],
+        ]
+        for inv_stat in inv_stats:
+            assert_raises_rpc_error(-8, 'Invalid requested statistic %s' % inv_sel_stat, node.getblockstats, height=1, stats=inv_stat)
+        # Make sure we aren't always returning inv_sel_stat as the culprit stat
+        assert_raises_rpc_error(-8, 'Invalid requested statistic aaa%s' % inv_sel_stat, node.getblockstats, height=1, stats=['minfee' , 'aaa%s' % inv_sel_stat])
+
+        # Make sure only the selected statistics are included
+        some_stats = {'minfee', 'maxfee'}
+        stats = node.getblockstats(height=1, stats=list(some_stats))
+        # Make sure only valid stats that have been requested appear
+        assert_equal(set(stats.keys()), some_stats)
+
```

Comment:
Suggest adding:

```
       # Make sure each stat can be queried on its own
       for stat in self.EXPECTED_STATS:
           for i in range(self.max_stat_pos+1):
               result = node.getblockstats(height=self.start_height + i, stats=[stat])
               assert_equal(list(result.keys()), [stat])
               if result[stat] != self.expected_stats[i][stat]:
                   self.log.info("result[%s] (%d) failed, %r != %r" % (stat, i, result[stat], self.expected_stats[i][stat]))
               assert_equal(result[stat], self.expected_stats[i][stat])
```

to ensure that each statistic can be queried on its own (in particular ensuring the GetBlockStatsMap() doesn't optimise out collecting the data needed for each stat). The above should automatically catch problems if new stats are introduced since it's just re-using the results from the all-the-stats check.